File: tools/source/sdks/java-sdk/examples/WorkerAuthorityQueryExample.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class WorkerAuthorityQueryExample {
    public static void main(String[] args) throws Exception {
        var client = new GameAgentEngineClient(
            System.getenv().getOrDefault("GAE_SERVER", "http://127.0.0.1:8080"),
            System.getenv().getOrDefault("GAE_KEY", "dev-key")
        );

        var dynamicInterfacesFile = System.getenv().getOrDefault(
            "GAE_DYNAMIC_INTERFACES_FILE",
            "tools/source/tests/runtime_task_dynamic_interfaces.json"
        );
        var dynamicInterfaces = Files.readString(Path.of(dynamicInterfacesFile));

        var worldId = System.getenv().getOrDefault("GAE_WORLD_ID", "demo_world");
        var nodeId = System.getenv().getOrDefault("GAE_NODE_ID", "innkeeper_001");
        var taskType = System.getenv().getOrDefault("GAE_TASK_TYPE", "npc_dialogue");
        var pipelineMode = System.getenv().getOrDefault("GAE_PIPELINE_MODE", "full");
        var message = System.getenv().getOrDefault(
            "GAE_MESSAGE",
            "Before answering, query the nearby scene state and then respond."
        );

        var requestJson = "{" +
            "\"world_id\":\"" + jsonEscape(worldId) + "\"," +
            "\"node_id\":\"" + jsonEscape(nodeId) + "\"," +
            "\"task_type\":\"" + jsonEscape(taskType) + "\"," +
            "\"messages\":[{" +
                "\"role\":\"user\"," +
                "\"content\":\"" + jsonEscape(message) + "\"" +
            "}]," +
            "\"context\":{" +
                "\"pipeline_mode\":\"" + jsonEscape(pipelineMode) + "\"," +
                "\"dynamic_interfaces\":" + dynamicInterfaces +
            "}" +
        "}";

        var response = client.invoke(requestJson);
        System.out.println(response);

        var callbackId = findFirstField(response, "callback_id");
        if (callbackId == null || callbackId.isBlank()) {
            System.out.println("No async data request was emitted.");
            return;
        }

        var tasksJson = client.listRuntimeTasks(null, null, 20);
        System.out.println(tasksJson);
        System.out.println("Find the runtime task with callback_id=" + callbackId);
        System.out.println("Next step: GameAgentWorker pull-once --consumer game_client");
    }

    private static String findFirstField(String json, String fieldName) {
        var pattern = Pattern.compile("\\\"" + Pattern.quote(fieldName) + "\\\"\\s*:\\s*\\\"([^\\\"]+)\\\"");
        Matcher matcher = pattern.matcher(json);
        if (matcher.find()) {
            return matcher.group(1);
        }
        return null;
    }

    private static String jsonEscape(String value) {
        return value
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\r", "\\r")
            .replace("\n", "\\n");
    }
}

File: tools/source/sdks/cs-sdk/examples/WorkerAuthorityQueryExample.cs
using System.Text.Json;
using GameAgentEngine.SDK;

var client = new GameAgentEngineClient(
    Environment.GetEnvironmentVariable("GAE_SERVER") ?? "http://127.0.0.1:8080",
    Environment.GetEnvironmentVariable("GAE_KEY") ?? "dev-key");

var dynamicInterfacesFile = Environment.GetEnvironmentVariable("GAE_DYNAMIC_INTERFACES_FILE")
    ?? "tools/source/workerhome/fixtures/runtime_task_dynamic_interfaces.json";
var dynamicInterfaces = JsonSerializer.Deserialize<List<DynamicInterface>>(
    await File.ReadAllTextAsync(dynamicInterfacesFile),
    new JsonSerializerOptions
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
    }) ?? new List<DynamicInterface>();

var response = await client.InvokeAsync(new InvokeRequest
{
    WorldId = Environment.GetEnvironmentVariable("GAE_WORLD_ID") ?? "demo_world",
    NodeId = Environment.GetEnvironmentVariable("GAE_NODE_ID") ?? "innkeeper_001",
    TaskType = Environment.GetEnvironmentVariable("GAE_TASK_TYPE") ?? "npc_dialogue",
    Messages = new List<ChatMessage>
    {
        new()
        {
            Role = "user",
            Content = Environment.GetEnvironmentVariable("GAE_MESSAGE") ?? "Before answering, query the nearby scene state and then respond."
        }
    },
    Context = new InvokeContext
    {
        PipelineMode = Environment.GetEnvironmentVariable("GAE_PIPELINE_MODE") ?? "full",
        DynamicInterfaces = dynamicInterfaces,
    }
});

Console.WriteLine($"request_id={response?.RequestId} task_type={response?.TaskType} execution_mode={response?.ExecutionMode}");
Console.WriteLine($"reply={response?.Reply ?? "<no reply>"}");

var dataRequest = response?.ActionCalls?.FirstOrDefault(item =>
    string.Equals(item.ActionId, "data_request", StringComparison.OrdinalIgnoreCase) &&
    string.Equals(item.Mode, "async", StringComparison.OrdinalIgnoreCase));

if (string.IsNullOrWhiteSpace(dataRequest?.CallbackId))
{
    Console.WriteLine("No async data request was emitted.");
    return;
}

var tasks = await client.ListRuntimeTasksAsync(limit: 20);
var task = tasks?.Tasks?.FirstOrDefault(item => item.CallbackId == dataRequest.CallbackId);
if (task is null)
{
    throw new InvalidOperationException($"Runtime task not found for callback_id={dataRequest.CallbackId}");
}

Console.WriteLine($"runtime_task_id={task.TaskId} interface={task.InterfaceName} consumer={task.Consumer} status={task.Status} callback_id={task.CallbackId}");
Console.WriteLine($"Next step: GameAgentWorker pull-once --consumer {task.Consumer ?? "game_client"}");
